### PR TITLE
cors: add unmatched_preflights

### PR DESCRIPTION
**Pull request by zirain**
xref: https://github.com/envoyproxy/envoy/pull/33051

Introduce `forward_not_matching_preflights` allow to return local reply when preflight origin does not match allowed origins.

### CODE REVIEW

File: networking/v1/virtual_service.proto [hunk 1]
```diff
@@ -1366,6 +1366,11 @@ message CorsPolicy {
   // (not the preflight) using credentials. Translates to
   // `Access-Control-Allow-Credentials` header.
   google.protobuf.BoolValue allow_credentials = 6;
+
+  // Indicates whether preflight requests not matching the configured 
+  // allowed origin should be forwarded to the upstream. 
+  // Default is true for compatibility.
```
Comment (hzxuzhonghu) [inline networking/v1/virtual_service.proto:1372]: should be false? It is a little wried to make a bool default value true

Comment (zirain) [reply networking/v1/virtual_service.proto:1372]: yes, should rename it and make it default false.

Comment (ramaraochavali) [reply networking/v1/virtual_service.proto]: Better to call it "donot_forward_unmatched_preflights" if we want to use Bool or use the original name but change it to bool like Envoy. Either of it works

**Review by ramaraochavali — APPROVED** [review]
LGTM

File: networking/v1/virtual_service.proto [hunk 2]
```diff
@@ -1366,6 +1366,11 @@ message CorsPolicy {
   // (not the preflight) using credentials. Translates to
   // `Access-Control-Allow-Credentials` header.
   google.protobuf.BoolValue allow_credentials = 6;
+
+  // Indicates whether preflight requests not matching the configured 
+  // allowed origin shouldn't be forwarded to the upstream. 
+  // Default is false.
+  google.protobuf.BoolValue donot_forward_unmatched_preflights = 8;
```
Comment (hzxuzhonghu) [inline networking/v1/virtual_service.proto:1373]: This is double negative, i donot think it is good. 

I mean  `forward_unmatched_preflights` default value false. 

Does this violate current behavior?

Comment (ramaraochavali) [reply networking/v1/virtual_service.proto:1373]: yes

Comment (zirain) [reply networking/v1/virtual_service.proto:1373]: if you want default false and keep current behavior, cannot avoid double negative.

Comment (howardjohn) [reply networking/v1/virtual_service.proto:1373]: You can make it an enum ?

`unmatched_preflights = UNSPECIFIED | FORWARD | IGNORE`

Comment (zirain) [reply networking/v1/virtual_service.proto:1373]: sounds good

File: networking/v1/virtual_service.proto [hunk 3]
```diff
@@ -1366,6 +1366,20 @@ message CorsPolicy {
...
+
+  // Indicates whether preflight requests not matching the configured 
+  // allowed origin shouldn't be forwarded to the upstream. 
+  // Default is forward to upstream.
+  UnmatchedPreflights unmatched_preflights = 8;
+
+  enum UnmatchedPreflights {
+    // Preflight requests not matching the configured allowed origin
+    // will be forwarded to the upstream.
+    FORWARD = 0;
```
Comment (howardjohn) [inline networking/v1/virtual_service.proto:1378]: I would set 0 to UNSPECIFIED. If we do `forward = 0`, we can never change our mind in the future and make 'ignore' the default. Even if we don't plan to now, its best to be precise

Comment (zirain) [reply networking/v1/virtual_service.proto]: updated

**Review by howardjohn — APPROVED** [review]

**Pull request merged**